typing/typemod.ml

```diff
@@ -1816,85 +1946,117 @@ let type_package env m p nl =
   (wrap_constraint env true modl mty Tmodtype_implicit, tl')
 
 (* Fill in the forward declarations *)
+
+let type_open ?used_slot ovf env loc me =
+  type_open_ ?used_slot ?toplevel:None ovf env loc me
+
+let gen_mod_ident me =
+  match me.pmod_desc with
+  | Pmod_functor _ | Pmod_extension _ | Pmod_ident _ -> None
+  | Pmod_structure _ | Pmod_apply _ | Pmod_constraint _ | Pmod_unpack _ ->
+      Some(gen_mod_ident ())
+
 let () =
   Typecore.type_module := type_module_alias;
   Typetexp.transl_modtype_longident := transl_modtype_longident;
   Typetexp.transl_modtype := transl_modtype;
-  Typecore.type_open := type_open_ ?toplevel:None;
+  Typecore.type_open := type_open;
   Typecore.type_package := type_package;
+  Typecore.gen_mod_ident := gen_mod_ident;
+  type_module_fwd := type_module;
   type_module_type_of_fwd := type_module_type_of
 
 
 (* Typecheck an implementation file *)
 
+let simplify_signature sg =
+  let rec aux = function
+    | [] -> [], StringSet.empty
+    | (Sig_value(id, _descr) as component) :: sg ->
+        let (sg, val_names) as k = aux sg in
+        let name = Ident.name id in
+        if StringSet.mem name val_names then k
+        else (component :: sg, StringSet.add name val_names)
+    | Sig_module({Ident.name}, _, _) :: sg
+        when open_generated name -> aux sg
+    | component :: sg ->
+        let (sg, val_names) = aux sg in
+        (component :: sg, val_names)
+  in
+  let (sg, _) = aux sg in
+  sg
+
 let type_implementation sourcefile outputprefix modulename initial_env ast =
-  Cmt_format.clear ();
-  try
-  Typecore.reset_delayed_checks ();
-  Env.reset_required_globals ();
-  if !Clflags.print_types then (* #7656 *)
-    Warnings.parse_options false "-32-34-37-38-60";
-  let (str, sg, finalenv) =
-    type_structure initial_env ast (Location.in_file sourcefile) in
-  let simple_sg = simplify_signature sg in
-  if !Clflags.print_types then begin
-    Typecore.force_delayed_checks ();
-    Printtyp.wrap_printing_env ~error:false initial_env
-      (fun () -> fprintf std_formatter "%a@." Printtyp.signature simple_sg);
-    (str, Tcoerce_none)   (* result is ignored by Compile.implementation *)
-  end else begin
-    let sourceintf =
-      Filename.remove_extension sourcefile ^ !Config.interface_suffix in
-    if Sys.file_exists sourceintf then begin
-      let intf_file =
-        try
-          find_in_path_uncap !Config.load_path (modulename ^ ".cmi")
-        with Not_found ->
-          raise(Error(Location.in_file sourcefile, Env.empty,
-                      Interface_not_compiled sourceintf)) in
-      let dclsig = Env.read_signature modulename intf_file in
-      let coercion =
-        Includemod.compunit initial_env ~mark:Includemod.Mark_positive
-          sourcefile sg intf_file dclsig
-      in
+  Misc.protect_refs [R(open_struct_level, 0); R(mod_ident_counter, 0)]
```

Comment:
The same should probably be done in Compile.interface and Optcompile.interface, no?